Личный проект: доверяй, но проверяй (часть 1)

German-D opened the pull request:
---
:mortar_board: [Личный проект: доверяй, но проверяй (часть 1)](https://up.htmlacademy.ru/javascript/19/user/831175/tasks/9)

:boom: https://htmlacademy-javascript.github.io/831175-kekstagram-19/4/

Suglobov commented:
В файле коммент с вопросом почему 2 раза вызывается. Потому что событие срабатывает сначала на лейблей, а потом на инпуте. 
Надо поставить условие, что event.target.tagName === 'INPUT' или event.target.tagName === 'LABEL'. Но лейбл больше помошник, поэтому можно поставить event.target.tagName !== 'LABEL', и это будет более менее универсальное условие.

Suglobov commented:
Опять есть глобальные переменные.
Так же надо навести порядок в объявлении и использовании переменных.
Раздели весь код по блокам. Потому что смотрю в код внизу, и мне приходится все время на верх мотать, чтоб посмотреть где и как объявились переменные.
В следующих уроках будет разделение, но можно уже сейчас некоторые блоки выделить.

Suglobov commented:
Во время поиска через регулярное выражение надо использовать флаги регулярки. Там есть флаг i, которые позволяет искать регистронезависимо, так же внутри регулярки в условиях прописаны как заглавные так и маленькие буквы, поэтому создавать массив с маленькими буквами нет необходимости.

Suglobov commented:
Проверять введенные данные в инпут надо не при вводе, а когда пользователь перестанет вводить.
Например ждать что прошло 500мс и тогда начать проверку.
А то проверка не простая и жрет ресурсы.

Suglobov commented:
hashtagsValue.split('# '); - в результате будет массив, в котором не будет символа # в начале никогда.
Split не включает в результат элементы, с помощью которых разбивает строку.
Правильно будет использовать регулярку /\s*/, тогда даже несколько пробельных символов будут удалены и не попадут в результат.

Suglobov commented:
Нашел твой бесконечный цикл.
В js примитивы при присваивании копируют свои значения, а массивы и объекты передают ссылку на себя. Правильно копировать массив надо var newArray = array.slice(); Тогда передадутся все значения, и это будут 2 разных массива.
Если сделать var newArray = array, то это 2 ссылки на один и тот же массив. newArray === array.
Теперь рассмотрим цикл в toLowCaseArray:
var lowCaseArray = array; - оба массива один одинаковее другого.
for (i = 0; i < array.length; i++) { - пока что массив имеет длинну 1
      lowCaseArray.push(array[i].toLowerCase()); - тут проиходит пуш и массив уже имеет длинну 2
    } - цикл теперь бесконечный.

German-D commented:
> В файле коммент с вопросом почему 2 раза вызывается. Потому что событие срабатывает сначала на лейблей, а потом на инпуте.
> Надо поставить условие, что event.target.tagName === 'INPUT' или event.target.tagName === 'LABEL'. Но лейбл больше помошник, поэтому можно поставить event.target.tagName !== 'LABEL', и это будет более менее универсальное условие.

Добавил проверку на LABEL

German-D commented:
> Опять есть глобальные переменные.
> Так же надо навести порядок в объявлении и использовании переменных.
> Раздели весь код по блокам. Потому что смотрю в код внизу, и мне приходится все время на верх мотать, чтоб посмотреть где и как объявились переменные.
> В следующих уроках будет разделение, но можно уже сейчас некоторые блоки выделить.

Убрал глобальную pinProgress внутрь функции

German-D commented:
> Нашел твой бесконечный цикл.
> В js примитивы при присваивании копируют свои значения, а массивы и объекты передают ссылку на себя. Правильно копировать массив надо var newArray = array.slice(); Тогда передадутся все значения, и это будут 2 разных массива.
> Если сделать var newArray = array, то это 2 ссылки на один и тот же массив. newArray === array.
> Теперь рассмотрим цикл в toLowCaseArray:
> var lowCaseArray = array; - оба массива один одинаковее другого.
> for (i = 0; i < array.length; i++) { - пока что массив имеет длинну 1
> lowCaseArray.push(array[i].toLowerCase()); - тут проиходит пуш и массив уже имеет длинну 2
> } - цикл теперь бесконечный.

Выпилил функцию лауркейса массива

German-D commented:
> hashtagsValue.split('# '); - в результате будет массив, в котором не будет символа # в начале никогда.
> Split не включает в результат элементы, с помощью которых разбивает строку.
> Правильно будет использовать регулярку /\s*/, тогда даже несколько пробельных символов будут удалены и не попадут в результат.

Переписал на var hashtagsArray = hashtagsValue.split('/\s*/');

German-D commented:
> Проверять введенные данные в инпут надо не при вводе, а когда пользователь перестанет вводить.
> Например ждать что прошло 500мс и тогда начать проверку.
> А то проверка не простая и жрет ресурсы.

пока решили не делать

Suglobov commented:
Для валидации не надо dom.imgUploadSubmit.setAttribute('disabled', 'disabled');
При нажатии на кнопку выйдет сообщение.

Suglobov commented:
} else {
      dom.imgUploadSubmit.removeAttribute('disabled');
      dom.textHashtags.setCustomValidity('');

    }
Тут else не нужен.

Suglobov commented:
var hashtagsArray = hashtagsValue.split(/\s/); - поставь /\s+/

Suglobov commented:
if (!uniqueArray.includes(array[i].toLowerCase())) {
        uniqueArray.push(array[i].toLowerCase());
      }
Тут 2 раза приводим строку в нижний регистр. Есть смысл задуматься над переменной.

German-D commented:
> Для валидации не надо dom.imgUploadSubmit.setAttribute('disabled', 'disabled');
> При нажатии на кнопку выйдет сообщение.


Убрал дизейблед

German-D commented:
> } else {
> dom.imgUploadSubmit.removeAttribute('disabled');
> dom.textHashtags.setCustomValidity('');
> 
> ```
> }
> ```
> 
> Тут else не нужен.

Убрал else

German-D commented:
> var hashtagsArray = hashtagsValue.split(/\s/); - поставь /\s+/

Поправил

German-D commented:
> if (!uniqueArray.includes(array[i].toLowerCase())) {
> uniqueArray.push(array[i].toLowerCase());
> }
> Тут 2 раза приводим строку в нижний регистр. Есть смысл задуматься над переменной.

Добавил переменную var arrayLowCase = array[i].toLowerCase();
